File: str_helper.cpp
// system includes
#include <map>

// includes
#include "str_helper.h"
#include "generic_protocol/str_helper.h"
#include "basic_objects/str_helper.h"
#include "basic_parser/str_helper.h"

namespace user_reg_protocol
{

namespace str_helper
{

using ::basic_parser::str_helper::write;
using ::basic_parser::str_helper::write_t;

// enums

#define TUPLE_VAL_STR(_x_)  _x_,#_x_

// objects

std::ostream & write( std::ostream & os, const User & r )
{
    os << "(";

    os << " gender="; ::basic_objects::str_helper::write( os, r.gender );
    os << " login="; write( os, r.login );
    os << " last_name="; write( os, r.last_name );
    os << " first_name="; write( os, r.first_name );
    os << " email="; ::basic_objects::str_helper::write( os, r.email );
    os << " phone="; write( os, r.phone );
    os << " birthday="; ::basic_objects::str_helper::write( os, r.birthday );

    os << ")";

    return os;
}

// base messages

std::ostream & write( std::ostream & os, const Request & r )
{
    os << "(";

    // no base class

    os << ")";

    return os;
}

std::ostream & write( std::ostream & os, const BackwardMessage & r )
{
    os << "(";

    // base class
    ::generic_protocol::str_helper::write( os, static_cast<const generic_protocol::BackwardMessage&>( r ) );


    os << ")";

    return os;
}

// messages

std::ostream & write( std::ostream & os, const RegisterUserRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const Request&>( r ) );

    os << " lead="; write( os, r.lead );

    return os;
}

std::ostream & write( std::ostream & os, const RegisterUserResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );


    return os;
}

std::ostream & write( std::ostream & os, const ConfirmRegistrationRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const Request&>( r ) );

    os << " registration_key="; write( os, r.registration_key );

    return os;
}

std::ostream & write( std::ostream & os, const ConfirmRegistrationResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );


    return os;
}

} // namespace str_helper

} // namespace user_reg_protocol

File: str_helper.php
<?php

namespace user_reg_protocol;


// base include
require_once __DIR__.'/../generic_protocol/str_helper.php';
// includes
require_once __DIR__.'/../basic_objects/str_helper.php';
require_once __DIR__.'/../basic_parser/str_helper.php';

// enums

// objects

function to_string__User( & $r )
{
    $res = "";

    $res .= "(";

    $res .= " gender=" . \basic_objects\to_string__gender_e( $r->gender );
    $res .= " login=" . \basic_parser\to_string__string( $r->login );
    $res .= " last_name=" . \basic_parser\to_string__string( $r->last_name );
    $res .= " first_name=" . \basic_parser\to_string__string( $r->first_name );
    $res .= " email=" . \basic_objects\to_string__Email( $r->email );
    $res .= " phone=" . \basic_parser\to_string__string( $r->phone );
    $res .= " birthday=" . \basic_objects\to_string__Date( $r->birthday );

    $res .= ")";

    return $res;
}

// base messages

function to_string__Request( & $r )
{
    $res = "";

    $res .= "(";

    // no base class

    $res .= ")";

    return $res;
}

function to_string__BackwardMessage( & $r )
{
    $res = "";

    $res .= "(";

    // base class
    $res .= \generic_protocol\to_string__BackwardMessage( $r );


    $res .= ")";

    return $res;
}

// messages

function to_string__RegisterUserRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__Request( $r );

    $res .= " lead=" . to_string__User( $r->lead );
    $res .= " password=" . \basic_parser\to_string__string( $r->password );

    return $res;
}

function to_string__RegisterUserResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );


    return $res;
}

function to_string__ConfirmRegistrationRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__Request( $r );

    $res .= " registration_key=" . \basic_parser\to_string__string( $r->registration_key );

    return $res;
}

function to_string__ConfirmRegistrationResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );


    return $res;
}

// generic

function to_string( $obj )
{
    $handler_map = array(
        // objects
        'user_reg_protocol\User'         => 'to_string__User',
        // messages
        'user_reg_protocol\RegisterUserRequest'         => 'to_string__RegisterUserRequest',
        'user_reg_protocol\RegisterUserResponse'         => 'to_string__RegisterUserResponse',
        'user_reg_protocol\ConfirmRegistrationRequest'         => 'to_string__ConfirmRegistrationRequest',
        'user_reg_protocol\ConfirmRegistrationResponse'         => 'to_string__ConfirmRegistrationResponse',
    );

    $type = get_class( $obj );

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\user_reg_protocol\\' . $handler_map[ $type ];
        return $func( $obj );
    }

    return \generic_protocol\to_string( $obj );
}

// namespace_end user_reg_protocol


?>
